histoqc/BaseImage.py

```diff
@@ -139,89 +92,200 @@ def addToPrintList(self, name, val):
+                    msg = f"Suggest using the smaller dimension thumbnail image because of the memory overhead."
+                    logging.warning(msg)
+                    self["warnings"].append(msg)
+                target_dims = getDimensionsByOneDim(self, int(size))
+                target_sampling_factor = img_base_size[0] / target_dims[0]
+                self[key] = getBestThumb(self, bx, by, target_dims, target_sampling_factor)
+
+        # can't determine operation.
+        else:
+            # TODO print out error message
+            err_msg = f"{self['filename']}: invalid arguments - {size}"
+            logging.error(err_msg)
+            self["warnings"].append(err_msg)
+            return
+        
         return self[key]
+
+def getBestThumb(s: BaseImage, x: int, y: int, dims: Tuple[int, int], target_sampling_factor: float):
+    osh = s["os_handle"]
+    
+    # get thumb from og
+    if not s["enable_bounding_box"]:
+        return np.array(osh.get_thumbnail(dims))
+    
+    (level, isExactLevel) = s.getBestLevelForDownsample(target_sampling_factor)
+    
+    # check if get the existing level
+    if isExactLevel:
+        tile = osh.read_region((x, y), level, dims)
+        return np.asarray(rgba2rgb(s, tile)) if np.shape(tile)[-1]==4 else np.asarray(tile)
+    # scale down the thumb img from the next high level
+    else:
+        return resizeTileDownward(s, dims, target_sampling_factor, level)
+        
+'''
+the followings are helper functions 
+'''
+def resizeTileDownward(self, dims, target_downsampling_factor, level):
+    osh = self["os_handle"]
+    (bx, by, bwidth, bheight) = self["img_bbox"]
+    end_x = bx + bwidth
+    end_y = by + bheight
+    
+    cloest_downsampling_factor = osh.level_downsamples[level]
+    win_size = 2048
+    
+    # create a new img
+    output = []
+    for x in range(bx, end_x, win_size):
+        row_piece = []
+        for y in range(by, end_y, win_size):
+            win_width, win_height = [win_size] * 2
+            # Adjust extraction size for endcut
+            if end_x < x + win_width:
+                win_width = end_x - x
+            if end_y < y +  win_height:
+                win_height = end_y - y
+
+            
+            win_down_width = int(round(win_width / target_downsampling_factor))
+            win_down_height = int(round(win_height / target_downsampling_factor))
+            
+            win_width = int(round(win_width / cloest_downsampling_factor))
+            win_height = int(round(win_height / cloest_downsampling_factor))
+            
+            cloest_region = osh.read_region((x, y), level, (win_width, win_height))
+            if np.shape(cloest_region)[-1]==4:
+                cloest_region = rgba2rgb(self, cloest_region)
+            target_region = cloest_region.resize((win_down_width, win_down_height))
+            row_piece.append(target_region)
+        row_piece = np.concatenate(row_piece, axis=0)
```

Comment:
can you adda comment here to note that this isn't very efficient, and if more efficiency isneeded we should likely refactor using "paste" from Image. or even just set the pixels directly with indexing?